File: PA1/Lex.java
               // Brian Pereira Alegria
// brpereir
// Lex.java



import java.io.*;
import java.util.Scanner;

public class Lex {
        public static void main(String[] args) throws IOException {
                Scanner in;
                PrintWriter out;
                String[] content;
                int n, m, k;
                int line;

                // check number of command line arguments is at least 2
                if (args.length != 2) {
                        System.err.println("Usage: input output");
                        System.exit(1);
                }
               // open files
                in = new Scanner(new File(args[0]));
                out = new PrintWriter(new FileWriter(args[1]));

                n = 0;

                // read lines from in
                while (in.hasNextLine()) {
                        in.nextLine();
                        n++;
                }

                content = new String[n];

                in.close();
                in = new Scanner(new File(args[0]));

                line = -1;

                // place content into array
                while (in.hasNextLine()) {
                        line++;
                        content[line] = in.nextLine();
                }

                List L = new List();
                L.append(0);
                k = content.length;

                // sort arreay accoording to value
                for (int i = 1; i < k; i++) {
                        m = i - 1;
                        L.moveFront();
                        while (0 <= m && content[i].compareTo(content[L.get()]) <= 0) {
                                L.moveNext();
                                m--;
                        }
                        if (0 > L.index()) {
                                L.append(i);
                        } else {
                                L.insertBefore(i);
                        }
                }

                // print content
                L.moveBack();
                for(int i = 0; i < k; i++){
                        out.println(content[L.get()]);
                        L.movePrev();
                }

                // close files
                in.close();
                out.close();
        }
}

File: PA2/Lex.c
  // Brian Pereira Alegria
// brpereir
// PA2
// Lex.c

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include"List.h"

#define MAX_LEN 1000

int main(int argc, char *argv[]) {

        int numLine;
        int n, m, k;
        FILE *in, *out;
        char line[MAX_LEN];
        // check that number of command line argument is at least 3
        if (argc != 3) {
                printf("Usage: %s <input file> <output file>\n", argv[0]);
                exit(1);
        }
  
  // open files
        in = fopen(argv[1], "r");
        out = fopen(argv[2], "w");
        if (in == NULL) {
                printf("Unable to open file %s for reading\n", argv[1]);
                exit(1);
        }
        if (out == NULL) {
                printf("Unable to open file %s for writing\n", argv[2]);
                exit(1);
        }
        numLine = 0;

        // read lines from in
        while (fgets(line, MAX_LEN, in) != NULL) {
                numLine++;
        }
        char *arr[numLine];
        n = 0;
        rewind(in); // set the file position to beginning, instead of closing & opening file again.
        while (fgets(line, MAX_LEN, in) != NULL) {
                arr[n] = malloc((strlen(line)) * sizeof(char *));
                strcpy(arr[n], line);// copy string
                n++;
        }

        List L = newList();
        append(L, 0);

        for (m = 1; m < numLine; m++) {
                moveFront(L);

                while (0 <= index(L)) {
                        if (strcmp(arr[m], arr[get(L)]) < 0) {
                                insertBefore(L, m);
                                break;
                        }
                        moveNext(L);
                }
                if (0 > index(L)) {
                        append(L, m);
                }
        }
        moveFront(L);

        while (0 <= index(L)) {
                fprintf(out, "%s", arr[get(L)]);
                moveNext(L);
        }

        k = 0;
        // close files
        fclose(in);
        fclose(out);
        // free memory
        while(k < m){
                free(arr[k]);
                k++;
        }
        freeList(&L);
}
